feat(api): Make RESTManager generic on RESTObject class

igorp-collabora opened the pull request:
Currently mixins like ListMixin are type hinted to return base RESTObject instead of a specific class like `MergeRequest`.

The GetMixin and GetWithoutIdMixin solve this problem by defining a new `get` method for every defined class. However, this creates a lot of duplicated code.

Make RESTManager use `typing.Generic` as its base class and use and assign the declared TypeVar to the `_obj_cls` attribute as a type of the passed class.

Make both `_obj_cls` and `_path` attributes an abstract properties so that type checkers can check that those attributes were properly defined in subclasses. Mypy will only check then the class is instantiated which makes non-final subclasses possible. Unfortunately pylint will check the declarations not instantiations so add `# pylint: disable=abstract-method` comments to all non-final subclasses like ListMixin.

Make `_path` attribute always be `str` instead of sometimes `None`. This eliminates unnecessary type checks.

Change all mixins like ListMixin or GetMixin to a subclass. This makes the attribute declarations much cleaner as for example `_list_filters` is now the only attribute defined by ListMixin.

Change SidekiqManager to not inherit from RESTManager and only copy its `__init__` method. This is because SidekiqManager never was a real manager and does not define `_path` or `_obj_cls`.

Delete `tests/unit/meta/test_ensure_type_hints.py` file as the `get` method is no required to be defined for every class.

Closes #2228 
Closes #3080

igorp-collabora commented:
This is a lot more verbose than #3080 but I think it is a better solution with less hacks and `# type: ignore[override]`

I did not make a RESTObjectList a generic or add overloads to list methods here to keep this PR more focused  on RESTManager but I want to do a follow-up PR in the future.

igorp-collabora commented:
I also have a 3rd implementation there I made RESTManager a Protocol as described in this comment

https://github.com/python-gitlab/python-gitlab/issues/2228#issuecomment-2578182800

It allows mixins to be "mixins" instead of subclasses using  `def method(sefl: RESTManager)`.

However, it quickly ran in to several issues:

1. Not possible to mixin attributes as `self` would only have attributes from RESTManager. You can move all mixins attributes to RESTManager or define new protocols that subclass the RESTManager but it pretty ugly. It seems like the ListMixins are always mixed with RESTManager so using subclasses instead of mixins is a cleaner solution.
2. Not possible to `issubclass(some_class, RESTManager)`. There is something about Protocols which prevent subclass checks.

igorp-collabora commented on tests/unit/meta/test_mro.py:
```diff
@@ -107,8 +108,13 @@ class definition.
+                if mro[index_to_check] == Generic:
+                    index_to_check -= 1
+
+                if mro[index_to_check].__module__ != "gitlab.base":
```
I added a work-around for the Generic class but I think this test is not required at all for the RESTManager and its mixins. Because the mixins are now subclasses the MRO will fail unless the RESTManager class is after all mixins. Meaning `class Name(gitlab.base.RESTManager, CreateMixin, UpdateMixin)` will fail and `class Name(CreateMixin, UpdateMixin, gitlab.base.RESTManager)` should be used instead.

nejch replied:
I agree since manager mixins are no longer mixins it's not needed there. Maybe we can get rid of this custom test later if it's done for all of them.

igorp-collabora commented:
> 1 PR to add native typing, which should already solve it for other methods, and another for removing the custom overrides and test

I can do that but I want to check if keeping the `def get` around would cause the type checking failure. This would make the commit non-atomic and I am not sure if you are ok with it. Alternatively the `get` removal commit can be first and then the generic commit after.

> I like the idea of defining a path on the `SidekiqManager` for consistent typing. We can actually make it `/sidekiq` and then use that for the methods using `{self.path}/endpoint`, which we already do in many cases:

I will add this.

nejch commented:
@igorp-collabora separate commits but the same PR would also work for me if that's ok! :bow:

igorp-collabora commented:
I rebased the PR and updated all the new RESTManager classes added.

I will split the `def get` removals to the separate commit tomorrow. I think I will place the get removal commit first followed by adding a generic type hints. The reverse order requires modifying all the get requests anyway because the redundant `cast` would need to be removed.

nejch reviewed (changes requested):
I managed to get through the diff and left some ramblings (some are non-blocking, just general notes).

In general I like that this will give users accurate typing for most resources.

It might feel a bit odd that our mixins are no longer mixins at all (not just in the typing context), but seems to be more maintainable than the protocol approach for now and the user value is more important :+1: We can revisit the protocol approach at some point if we want the code closer to the user (`gitlab.v4.objects.*`) to be less type-verbose but IMO this is OK.

Not sure if @JohnVillalovos would also like to give this another look as well :bow:

igorp-collabora commented on gitlab/v4/cli.py:
```diff
@@ -28,25 +28,18 @@ def __init__(
         # the class _path attribute, and replace the value with the result.
         if TYPE_CHECKING:
-            assert self.mgr_cls._path is not None
+            assert isinstance(self.mgr_cls._path, str)
```
There is an issue with the `@abstractmethod @property` where type checker will only consider `_path` an `str` if it is called from an object but then called from class it will be considered a `Callble[[RESTManager], str]`.

However, I think I will change `_path` from an abstract property to a simple type annotation. The downside will be the lack of type checking that the class was defined properly but it will keep the type consistent between class and object.

igorp-collabora commented on gitlab/v4/cli.py:
```diff
@@ -28,25 +28,18 @@ def __init__(
-            Type[gitlab.mixins.ListMixin],
-            Type[gitlab.mixins.UpdateMixin],
-        ] = getattr(gitlab.v4.objects, f"{self.cls.__name__}Manager")
+        self.mgr_cls: Any = getattr(gitlab.v4.objects, f"{self.cls.__name__}Manager")
```
I tried to make it type check properly but it turned out to require too many changes. I can try making the old annotation work.

nejch replied:
OK, let's leave it, it's not in the API.

igorp-collabora commented on gitlab/mixins.py:
```diff
@@ -986,13 +939,19 @@ def participants(
     )
     @exc.on_http_error(exc.GitlabRenderError)
-    def render(self, link_url: str, image_url: str, **kwargs: Any) -> Dict[str, Any]:
+    def render(
+        self,
+        link_url: str,
+        image_url: str,
+        **kwargs: Any,
+    ) -> Dict[str, Any]:
```
Most likely this is because an earlier attempt at using `self: RESTManagerProtocol`. It resulted in line being longer than 88 so black would expand it. I will revert this.

igorp-collabora commented on gitlab/v4/objects/clusters.py:
```diff
@@ -56,19 +56,14 @@ def create(
-        self, id: Union[str, int], lazy: bool = False, **kwargs: Any
-    ) -> GroupCluster:
-        return cast(GroupCluster, super().get(id=id, lazy=lazy, **kwargs))
+        return super().create(data, path=path, **kwargs)
```
Yes, mypy complained about redundant cast.

igorp-collabora commented on gitlab/v4/objects/sidekiq.py:
```diff
@@ -33,7 +36,7 @@ def queue_metrics(self, **kwargs: Any) -> Union[Dict[str, Any], requests.Respons
             Information about the Sidekiq queues
         """
-        return self.gitlab.http_get("/sidekiq/queue_metrics", **kwargs)
+        return self.gitlab.http_get(f"{self._path}/queue_metrics", **kwargs)
```
There shouldn't be any difference because the `"/sidekiq"` does not have `{}` formatting specifiers. However,  `self.path` is more consistent. I will fix this.

igorp-collabora commented on gitlab/v4/objects/users.py:
```diff
@@ -525,46 +511,45 @@ class UserGPGKey(ObjectDeleteMixin, RESTObject):
     """
 
     _path = "/users/{user_id}/identities"
+    _obj_cls = User
```
Yes, same work-around as `Sidekiq` can work here.

Alternatively, it should be possible to split `RESTManager` in to two classes. `RESTManagerBase` that only defines the `_path` and `RESTManager` that inherits base class and adds the `_obj_cls`. This also will eliminate the need for the work-arounds in the `gitlab/v4/cli.py`.

nejch replied:
@igorp-collabora that might work well if it means we get rid of a lot of boilerplate yes!

igorp-collabora replied:
@nejch I added the unit test to verify the `_obj_cls` and `_path` arguments.

However, after looking how CLI is implemented there is an issue with how the endpoint discovery works and splitting the RESTManager in to the base class.

Right now the discovery does a cycle of `RESTManager` -> obtain _obj_cls -> `RESTObject` -> `str(RESTObject.__class__.__name__) + "Manager"` -> `RESTManager` for some reason. This won't work if the RESTManager does not have a `_obj_cls` Not sure why it can't just iterate over `RESTManager`.

Removing this discovery cycle can be done in a separated PR. Otherwise I think this PR is already large enough.

igorp-collabora replied:
This also means the Sidekiq does not show up in CLI in the current version:

```
    runner-job          API endpoint: /runners/{runner_id}/jobs
    shared-project      API endpoint:
                        /groups/{group_id}/projects/shared
    snippet             API endpoint: /snippets
    starred-project     API endpoint:
                        /users/{user_id}/starred_projects
    todo                API endpoint: /todos
```

nejch replied:
Sorry for wasting time here, the CLI is mostly legacy code but still widely used. I think this is sufficient for now :bow:

igorp-collabora replied:
No problem. This just means there is room for another PR to improve CLI endpoint discovery.

nejch commented on gitlab/base.py:
```diff
@@ -346,16 +360,17 @@ class RESTManager:
     _parent_attrs: Dict[str, Any]
     gitlab: Gitlab
 
+    _path: ClassVar[str]
+    _obj_cls: type[TObjCls]
```
These 2 could go back to lines 349-350 to reduce the diff.

igorp-collabora replied:
Done. The diff is now:

```diff
-    _path: Optional[str] = None
-    _obj_cls: Optional[Type[RESTObject]] = None
+    _path: ClassVar[str]
+    _obj_cls: type[TObjCls]
```

nejch commented on gitlab/mixins.py:
```diff
@@ -73,30 +70,23 @@ def head(
     def get(
-        self, id: Union[str, int], lazy: bool = False, **kwargs: Any
-    ) -> base.RESTObject:
+        self,
+        id: Union[str, int],
+        lazy: bool = False,
+        **kwargs: Any,
+    ) -> base.TObjCls:
```
I must have missed these before. There's a few more unrelated formatting changes (im just nitpickign here because it's a large PR so any change to make the diff smaller helps :sweat_smile: )

```suggestion
        self, id: Union[str, int], lazy: bool = False, **kwargs: Any
    ) -> base.TObjCls:
```

nejch commented on gitlab/mixins.py:
```diff
@@ -128,18 +116,14 @@ def get(
 
     @exc.on_http_error(exc.GitlabGetError)
-    def get(self, **kwargs: Any) -> base.RESTObject:
+    def get(
+        self,
+        **kwargs: Any,
+    ) -> base.TObjCls:
```
```suggestion
    def get(self, **kwargs: Any) -> base.TObjCls:
```

nejch commented on gitlab/mixins.py:
```diff
@@ -194,18 +175,14 @@ def refresh(self, **kwargs: Any) -> None:
 
     @exc.on_http_error(exc.GitlabListError)
-    def list(self, **kwargs: Any) -> Union[base.RESTObjectList, List[base.RESTObject]]:
+    def list(
+        self,
+        **kwargs: Any,
+    ) -> Union[base.RESTObjectList, List[base.TObjCls]]:
```
```suggestion
    def list(self, **kwargs: Any) -> Union[base.RESTObjectList, List[base.TObjCls]]:
```

nejch commented on gitlab/mixins.py:
```diff
@@ -244,37 +221,22 @@ def list(self, **kwargs: Any) -> Union[base.RESTObjectList, List[base.RESTObject
     def create(
-        self, data: Optional[Dict[str, Any]] = None, **kwargs: Any
-    ) -> base.RESTObject:
+        self,
+        data: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ) -> base.TObjCls:
```
```suggestion
        self, data: Optional[Dict[str, Any]] = None, **kwargs: Any
    ) -> base.TObjCls:
```

nejch commented on gitlab/mixins.py:
```diff
@@ -384,17 +339,14 @@ def update(
+class SetMixin(base.RESTManager[base.TObjCls]):
     @exc.on_http_error(exc.GitlabSetError)
-    def set(self, key: str, value: str, **kwargs: Any) -> base.RESTObject:
+    def set(
+        self,
+        key: str,
+        value: str,
+        **kwargs: Any,
+    ) -> base.TObjCls:
```
```suggestion
    def set(self, key: str, value: str, **kwargs: Any) -> base.TObjCls:
```

nejch commented on gitlab/mixins.py:
```diff
@@ -414,21 +366,16 @@ def set(self, key: str, value: str, **kwargs: Any) -> base.RESTObject:
+class DeleteMixin(base.RESTManager[base.TObjCls]):
     @exc.on_http_error(exc.GitlabDeleteError)
-    def delete(self, id: Optional[Union[str, int]] = None, **kwargs: Any) -> None:
+    def delete(
+        self: base.RESTManager[Any],
+        id: Optional[Union[str, int]] = None,
+        **kwargs: Any,
+    ) -> None:
```
```suggestion
    def delete(
        self: base.RESTManager[Any], id: Optional[Union[str, int]] = None, **kwargs: Any
    ) -> None:
```

Related question: do we need the explicit self annotation here? If not, it can probably go back to the one-liner.

nejch commented on gitlab/mixins.py:
```diff
@@ -708,7 +642,10 @@ class RotateMixin(_RestManagerBase):
     @exc.on_http_error(exc.GitlabRotateError)
     def rotate(
-        self, id: Union[str, int], expires_at: Optional[str] = None, **kwargs: Any
+        self,
+        id: Union[str, int],
+        expires_at: Optional[str] = None,
+        **kwargs: Any,
```
```suggestion
        self, id: Union[str, int], expires_at: Optional[str] = None, **kwargs: Any
```

nejch reviewed (changes requested):
> @nejch I think this PR is in good enough form. The only possible change I might add is splitting the `RESTManager` in to 2 classes. (the parent one with just `_path` and one that adds `_obj_cls`)
> 
> I think there is also room to add a new unit test that verifies that `_obj_cls` and `_path` are properly set for subclasses. Since they are no longer defined using `@abstractmethod @property` the type checkers will not verify that they properly defined.

@igorp-collabora agree, I think this is almost good to go! Now that I can focus just on a few files I just found a few more nits that I forgot earlier. I think the base manager could work well if it helps with the code using it. Let me know if that's ok.

I'm happy to see typing and linting here helps get rid of our custom checks so if possible I'd prefer to avoid more of them - maybe we can have a follow-up for that though.

JohnVillalovos commented:
@nejch I'm not sure if I will have time to review this in the near future. I trust your reviews though 👍🤓

JohnVillalovos commented:
@nejch After this gets merged, I will propose a PR to update the black config so that it will do the equivalent of `--skip-magic-trailing-comma`.

Basically add to  `pyproject.toml`:

```
[tool.black]
skip-magic-trailing-comma = true
```

igorp-collabora commented:
I made a separate PR to remove the `get` methods: #3110 

Once it is merged I will rebase this PR. It should shrink by quite a bit after that.

igorp-collabora commented:
Rebase done.

The line diff is now `+748 -536`. Before it was `+765 -1485`.

nejch reviewed (approved):
This LGTM, I'll just wait with merging until later today so I can do a checkout on some of our projects that make heavy use of the library. Also it'll be daytime in Oregon by then in case @JohnVillalovos has any objections :grin: 
This should have no impact at runtime but if mypy starts complaining a lot, it might be good if we add a breaking change so people can migrate more easily when they get their dependency/renovate bot bumps.

Thanks a lot for your patience @igorp-collabora and for all the effort put in here!

igorp-collabora commented:
I think there shouldn't be any difference with the Get methods because redefined methods, however, other mixins might make certain code no longer pass type checking. Biggest changes that might impact type checking:

1. `RESTObject.path` is now always `str`. This means checking for None is no longer required. I don't think mypy checks for unreachable code but there might be some redundant casts.
2. `ListMixin` now returns a list of concrete subclasses as a part of union. I also want to make the `RESTObjectList` generic in the future and add overloads just like download methods.
3. `CreateMixin` returns concrete subclass.
4. `SetMixin` returns concrete subclass.

JohnVillalovos commented:
> Also it'll be daytime in Oregon by then in case @JohnVillalovos has any objections 😁

No objections from me. Sorry I haven't had the time to review this yet. I trust your decision @nejch

igorp-collabora commented:
Once this gets merged I plan on adding more typing changes:

* Make `RESTObjectList` generic to type hint iterated object class
* Add overloads to the list methods so that only `RESTObjectList` or `list` gets returned based on the arguments. This can also make skipping `get_all` have a deprecation warning then type checked.

nejch commented:
> Once this gets merged I plan on adding more typing changes:
> 
>     * Make `RESTObjectList` generic to type hint iterated object class
> 
>     * Add overloads to the list methods so that only `RESTObjectList` or `list` gets returned based on the arguments. This can also make skipping `get_all` have a deprecation warning then type checked.

Thanks @igorp-collabora. I checked out python-gitlab from your branch in a few of our projects and generally it all looks great but indeed the list methods are now failing checks where we previously had workarounds for the union types. I think just for clarity I'd make that a breaking change, but we can do this in a follow-up since the `list()` return types will change with the incoming PRs again.

I'll go ahead and merge this for now!
